File: src/core/services/catalog/datacatProvider.js
/**
 * DataCat (datacat.run) provider.
 * Anonymous session via deviceToken → sessionToken.
 * All API calls go through catalogHttp (CapacitorHttp on native, corsproxy on web).
 */
import { catalogGet, catalogPost } from './catalogHttp.js';

const BASE = 'https://datacat.run';
const KEY_DEVICE = 'gz_dc_device';
const KEY_TOKEN = 'gz_dc_token';

// ─── Session ──────────────────────────────────────────────────────────────────

function getDeviceToken() {
    let token = localStorage.getItem(KEY_DEVICE);
    if (!token) {
        // Generate a UUID v4
        token = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
            const r = Math.random() * 16 | 0;
            return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
        });
        localStorage.setItem(KEY_DEVICE, token);
    }
    return token;
}

function getSessionToken() {
    return localStorage.getItem(KEY_TOKEN);
}

function setSessionToken(token) {
    localStorage.setItem(KEY_TOKEN, token);
}

/**
 * Initialize anonymous DataCat session. Stores sessionToken in localStorage.
 * @returns {Promise<string>} sessionToken
 */
export async function datacatInit() {
    const deviceToken = getDeviceToken();
    const data = await catalogPost(`${BASE}/api/liberator/identify`, { deviceToken }, {
        'Origin': 'https://datacat.run',
        'Referer': 'https://datacat.run/'
    });
    if (!data?.sessionToken) throw new Error('DataCat: no sessionToken in response');
    setSessionToken(data.sessionToken);
    return data.sessionToken;
}

/**
 * Returns a valid sessionToken, initializing if needed.
 */
async function getToken() {
    let token = getSessionToken();
    if (!token) token = await datacatInit();
    return token;
}

function authHeaders(token) {
    return {
        'X-Session-Token': token,
        'Origin': 'https://datacat.run',
        'Referer': 'https://datacat.run/'
    };
}

/**
 * Validate current session token.
 * @returns {Promise<boolean>}
 */
export async function datacatValidate() {
    const token = getSessionToken();
    if (!token) return false;
    try {
        await catalogGet(`${BASE}/api/characters/recent-public?limit=1&summary=1`, authHeaders(token));
        return true;
    } catch (e) {
        if (e.status === 401 || e.status === 403) {
            localStorage.removeItem(KEY_TOKEN);
            return false;
        }
        return true; // network error — assume still valid
    }
}

/**
 * Ensure session is valid; re-init if not.
 */
export async function datacatEnsureSession() {
    const valid = await datacatValidate();
    if (!valid) await datacatInit();
}

// ─── Browse / Search ──────────────────────────────────────────────────────────

const MIN_TOKENS = 889;

/**
 * Browse recent public characters.
 * @param {{ page?: number, limit?: number, tagIds?: number[], nsfw?: boolean }} opts
 */
export async function datacatBrowse({ page = 1, limit = 24, tagIds = [], nsfw = false, filters = {} } = {}) {
    const token = await getToken();
    const offset = (page - 1) * limit;
    const minTok = filters.minTokens ?? MIN_TOKENS;
    const isNsfw = filters.nsfw !== undefined ? filters.nsfw : nsfw;
    const filterTagIds = filters.tagIds?.length ? filters.tagIds : tagIds;

    const params = new URLSearchParams({
        limit: String(limit),
        offset: String(offset),
        summary: '1',
        minTotalTokens: String(minTok)
    });
    if (filters.maxTokens) params.set('maxTotalTokens', String(filters.maxTokens));

    if (filterTagIds.length) params.set('tagIds', filterTagIds.join(','));
    if (!isNsfw) params.set('nsfw', '0');

    // If there's a sort like "oldest", DataCat might support sortBy=oldest, but recent-public defaults to newest.

    const data = await catalogGet(`${BASE}/api/characters/recent-public?${params}`, authHeaders(token));
    return {
        characters: (data.characters || []).map(normalizeListItem),
        total: data.totalCount || 0
    };
}

/**
 * Get trending/fresh characters.
 * @returns {{ last24h: CatalogItem[], thisWeek: CatalogItem[] }}
 */
export async function datacatFresh() {
    const token = await getToken();
    const data = await catalogGet(
        `${BASE}/api/characters/fresh?summary=1&sortBy=score&limit24=20&limitWeek=20`,
        authHeaders(token)
    );
    return {
        last24h: (data.windows?.last24h?.characters || []).map(normalizeListItem),
        thisWeek: (data.windows?.thisWeek?.characters || []).map(normalizeListItem)
    };
}

/**
 * Search characters by query.
 */
export async function datacatSearch({ query, page = 1, limit = 24, filters = {} } = {}) {
    const token = await getToken();
    const offset = (page - 1) * limit;
    const minTok = filters.minTokens ?? MIN_TOKENS;
    const isNsfw = filters.nsfw !== undefined ? filters.nsfw : false;
    const filterTagIds = filters.tagIds || [];

    const params = new URLSearchParams({
        limit: String(limit),
        offset: String(offset),
        summary: '1',
        minTotalTokens: String(minTok)
    });
    if (filters.maxTokens) params.set('maxTotalTokens', String(filters.maxTokens));
    if (!isNsfw) params.set('nsfw', '0');
    if (filterTagIds.length) params.set('tagIds', filterTagIds.join(','));
    if (query) params.set('search', query);

    const data = await catalogGet(`${BASE}/api/characters/recent-public?${params}`, authHeaders(token));
    return {
        characters: (data.characters || []).map(normalizeListItem),
        total: data.totalCount || 0
    };
}

// ─── Character Download ───────────────────────────────────────────────────────

/**
 * Download a character card in V2-like format and convert to Glaze schema.
 * @param {string} uuid
 * @returns {Promise<{ charData: object, avatarUrl: string }>}
 */
export async function datacatGetCharacter(uuid) {
    const token = await getToken();
    const ts = Date.now();
    const data = await catalogGet(
        `${BASE}/api/characters/${uuid}/download?t=${ts}`,
        authHeaders(token)
    );

    const raw = data.data || data;
    return {
        charData: convertToGlaze(raw),
        avatarUrl: resolveAvatarUrl(raw.avatar)
    };
}

// ─── JanitorAI Extraction ─────────────────────────────────────────────────────

const IDEMPOTENCY_KEY = () => 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
    const r = Math.random() * 16 | 0;
    return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
});

/**
 * Submit a JanitorAI character URL for cloud-browser extraction via DataCat.
 * @param {string} janitorUrl e.g. https://janitorai.com/characters/{uuid}
 * @param {boolean} publicFeed Whether to appear on DataCat public feed
 * @returns {Promise<{ queued: boolean, started: boolean, queuePosition: number }>}
 */
export async function datacatExtract(janitorUrl, publicFeed = true) {
    const token = await getToken();
    return catalogPost(
        `${BASE}/api/character/smart-extract-v2`,
        {
            url: janitorUrl,
            appearOnPublicFeed: publicFeed,
            useSeparateWorkerServer: true,
            inlinePostExtractCreatorProfile: true,
            idempotencyKey: IDEMPOTENCY_KEY()
        },
        authHeaders(token)
    );
}

/**
 * Poll extraction status.
 * @returns {Promise<{ inProgress: object|null, queue: object[], history: object[] }>}
 */
export async function datacatExtractionStatus() {
    const token = await getToken();
    const ts = Date.now();
    const data = await catalogGet(`${BASE}/api/extraction/status?t=${ts}`, authHeaders(token));
    return {
        inProgress: data.inProgress || null,
        queue: data.queue || [],
        history: data.history || []
    };
}

// ─── Normalization ────────────────────────────────────────────────────────────

function stripEmoji(str) {
    if (!str) return str;
    return str.replace(/[\u{1F300}-\u{1FFFF}\u{2600}-\u{27BF}\uFE0F\u200D\s]+/gu, '').trim();
}

const IMAGE_BASE = 'https://ella.janitorai.com/bot-avatars/';

function resolveAvatarUrl(url) {
    if (!url) return null;
    if (url.startsWith('http')) return url;
    return IMAGE_BASE + url;
}

function normalizeListItem(c) {
    return {
        id: c.characterId || c.character_id || c.uuid || c.id,
        name: c.chatName || c.chat_name || c.name || 'Unknown',
        avatarUrl: resolveAvatarUrl(c.avatar),
        tags: (c.tags || []).map(t => (typeof t === 'string' ? stripEmoji(t) : stripEmoji(t.name))).filter(Boolean),
        tokens: c.totalTokens || c.total_tokens || 0,
        creator: c.creatorName || c.creator_name || '',
        nsfw: Boolean(c.is_nsfw),
        source: 'datacat'
    };
}

function convertToGlaze(raw) {
    const tags = (raw.tags || [])
        .map(t => (typeof t === 'string' ? stripEmoji(t) : stripEmoji(t?.name)))
        .filter(Boolean);

    return {
        name: raw.name || raw.chatName || raw.chat_name || 'Unknown',
        description: raw.personality || raw.description || '',
        personality: '',
        scenario: raw.scenario || '',
        first_mes: raw.first_mes || raw.first_message || '',
        mes_example: raw.mes_example || '',
        creator_notes: raw.description || raw.creator_notes || '',
        system_prompt: raw.system_prompt || '',
        post_history_instructions: raw.post_history_instructions || '',
        alternate_greetings: Array.isArray(raw.alternate_greetings) ? raw.alternate_greetings : [],
        tags,
        creator: raw.creator || '',
        character_book: raw.character_book || null,
        extensions: { datacat: { id: raw.characterId || raw.character_id || raw.uuid || raw.id } }
    };
}

File: lib/features/catalog/services/datacat_provider.dart
import 'dart:math';

import 'package:dio/dio.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'greeting_normalizer.dart';
import 'catalog_http.dart';
import '../catalog_models.dart';
import 'extraction_status.dart';

/// URL extraction: asking DataCat to go and index a character it has never
/// seen, then reading what came out.
///
/// Everything else DataCat is used for — browsing, search, tags, creators,
/// community, and the card/image transfer — now runs on the official Client
/// API under `services/datacat/`. This file is what could not move: the Client
/// API is read-mostly over DataCat's existing index, and has no endpoint that
/// takes a URL and produces a new row. So the undocumented site endpoints, the
/// anonymous session token they need, and the source-shaped row reader below
/// stay, scoped to the one job that still requires them.
///
/// Two features depend on it: importing a character by pasting its URL, and
/// recovering a JanitorAI card whose definition its creator closed (DataCat
/// scrapes those, and the public endpoint does not serve them).

const _base = 'https://datacat.run';
const _keyDevice = 'gz_dc_device';
const _keyToken = 'gz_dc_token';
const _saucepanCdnBase = 'https://cdn.saucepan.ai';
const _imageBase = 'https://ella.janitorai.com/bot-avatars/';

String _uuid() {
  final r = Random();
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replaceAllMapped(
    RegExp('[xy]'),
    (m) {
      final v = r.nextInt(16);
      return (m.group(0) == 'x' ? v : (v & 0x3 | 0x8)).toRadixString(16);
    },
  );
}

Future<String> _getDeviceToken() async {
  final prefs = await SharedPreferences.getInstance();
  var token = prefs.getString(_keyDevice);
  if (token == null) {
    token = _uuid();
    await prefs.setString(_keyDevice, token);
  }
  return token;
}

Future<String?> _getSessionToken() async {
  final prefs = await SharedPreferences.getInstance();
  return prefs.getString(_keyToken);
}

Future<void> _setSessionToken(String token) async {
  final prefs = await SharedPreferences.getInstance();
  await prefs.setString(_keyToken, token);
}

Map<String, String> _authHeaders(String token) => {
      'X-Session-Token': token,
      'Origin': 'https://datacat.run',
      'Referer': 'https://datacat.run/',
    };

Future<String> datacatInit() async {
  final deviceToken = await _getDeviceToken();
  final data = await catalogPost(
    '$_base/api/liberator/identify',
    {'deviceToken': deviceToken},
    {
      'Origin': 'https://datacat.run',
      'Referer': 'https://datacat.run/',
    },
  );
  final sessionToken = data['sessionToken'] as String?;
  if (sessionToken == null) throw Exception('DataCat: no sessionToken');
  await _setSessionToken(sessionToken);
  return sessionToken;
}

Future<String> _getToken() async {
  var token = await _getSessionToken();
  token ??= await datacatInit();
  return token;
}

/// Runs an authenticated DataCat call, re-establishing the session once when
/// the server rejects the token it was made with.
///
/// The anonymous session token comes from `/api/liberator/identify` and is kept
/// in SharedPreferences with no expiry, so it outlives whatever the server is
/// still willing to honour. A token DataCat has forgotten answers 401/403 to
/// every call carrying it, and the browse path used to be the only one that
/// knew how to replace one — it ran a throwaway probe request before each
/// search purely to find out. Every other endpoint dead-ended: the card detail
/// showed `HTTP 403: Forbidden` behind a Retry button that re-sent the same
/// dead token, which is why retrying never helped.
///
/// A 403 can also be DataCat's bot protection, which no amount of fresh
/// sessions will cure. One retry is what tells the two apart — and the wasted
/// request only happens on a failure that was already fatal.
Future<T> _datacatAuthed<T>(Future<T> Function(String token) call) async {
  final token = await _getToken();
  try {
    return await call(token);
  } on DioException catch (e) {
    final status = catalogErrorStatus(e);
    if (status != 401 && status != 403) rethrow;
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_keyToken);
    return await call(await datacatInit());
  }
}

Future<Map<String, dynamic>> _datacatGet(String path) =>
    _datacatAuthed((token) => catalogGet('$_base$path', _authHeaders(token)));

Future<Map<String, dynamic>> _datacatPost(
  String path,
  Map<String, dynamic> body,
) => _datacatAuthed(
  (token) => catalogPost('$_base$path', body, _authHeaders(token)),
);

String? _pickAvatarSource(Map<String, dynamic> raw, Map<String, dynamic> meta) {
  return (raw['avatar'] ?? raw['image'] ?? raw['image_url'] ??
          raw['avatar_url'] ?? raw['max_res_url'] ?? meta['image'] ??
          meta['image_url'] ?? meta['avatar'] ?? meta['avatar_url']) as String?;
}

String? _resolveAvatarUrl(String? url) {
  if (url == null) return null;
  if (url.startsWith('http')) return url;
  if (url.startsWith('//')) return 'https:$url';
  if (url.startsWith('/images/')) return '$_saucepanCdnBase$url';
  if (url.startsWith('images/')) return '$_saucepanCdnBase/$url';
  if (RegExp(r'^[0-9a-f-]+/highres$', caseSensitive: false).hasMatch(url)) {
    return '$_saucepanCdnBase/images/$url';
  }
  if (url.startsWith('/')) return 'https://ella.janitorai.com$url';
  if (!url.contains('/')) return '$_imageBase$url';
  return 'https://ella.janitorai.com/$url';
}

/// The active, non-placeholder primary content variant of a DataCat row. For
/// Saucepan cards with a hidden definition, DataCat's "Character Repair" job
/// exposes the recovered body here (with `description` overloaded to carry it).
Map<String, dynamic>? _pickRecoveryVariant(Map<String, dynamic> char) {
  final variants = char['content_variants'];
  if (variants is! List) return null;
  for (final v in variants) {
    if (v is Map && v['isPrimary'] == true && v['isRecoveryPlaceholder'] != true) {
      final content = v['content'];
      if (content is Map) return content.cast<String, dynamic>();
    }
  }
  return null;
}

/// Recovery-sourced `chara_card_v2_json` bodies carry edge
/// `##DESCRIPTION START##`-style delimiter lines; strip them.
String _stripDatacatMarkers(String text) {
  if (text.isEmpty) return text;
  return text
      .replaceFirst(RegExp(r'^\s*##[A-Z _]*(?:START|END)##[ \t]*\r?\n?'), '')
      .replaceFirst(RegExp(r'\r?\n?[ \t]*##[A-Z _]*(?:START|END)##\s*$'), '')
      .trim();
}

/// Maps a DataCat `/api/characters/{id}` row into a [CharacterData].
///
/// Field placement is source-dependent, ported from `buildV2FromDatacat` in the
/// SillyTavern-CharacterLibrary reference. For JanitorAI-source rows the real
/// definition is in `personality` and the website blurb (often HTML) is in
/// `description`; Saucepan rows overload `description` / recovery variants. We
/// keep the definition in [CharacterData.description] and the blurb in
/// [CharacterData.creatorNotes] — the same placement `buildV2FromDatacat` uses
/// (definition into `data.description`, `data.personality` left empty), so
/// `{{description}}` resolves to the prompt body and the blurb never lands in
/// it.
///
/// Only extraction results go through this. A character the Client API can
/// serve arrives as a standard Character Card V2 and is read by
/// `datacat/datacat_cards.dart`, which needs none of this guesswork.
///
/// Public so the field mapping can be tested against a row rather than
/// only through a live request: which greeting field a row carries is
/// exactly what #98 turned on.
CharacterData datacatCharacterData(Map<String, dynamic> char) {
  String s(dynamic v) => v == null ? '' : v.toString();
  String pick(List<String> xs) =>
      xs.firstWhere((e) => e.trim().isNotEmpty, orElse: () => '');

  final isSaucepan = char['primary_content_source_kind'] == 'saucepan';
  final recovered = _pickRecoveryVariant(char);
  final v2Data = (char['chara_card_v2_json'] is Map &&
          (char['chara_card_v2_json'] as Map)['data'] is Map)
      ? ((char['chara_card_v2_json'] as Map)['data'] as Map).cast<String, dynamic>()
      : const <String, dynamic>{};
  final companion = (char['companion_snapshot'] is Map)
      ? (char['companion_snapshot'] as Map).cast<String, dynamic>()
      : const <String, dynamic>{};

  final definition = isSaucepan
      ? pick([
          s(recovered?['description']), s(recovered?['personality']),
          s(v2Data['description']), s(char['description']),
        ])
      : pick([
          s(char['personality']), s(recovered?['personality']),
          _stripDatacatMarkers(s(v2Data['description'])),
        ]);
  final scenario =
      pick([s(char['scenario']), s(recovered?['scenario']), s(v2Data['scenario'])]);
  final creatorNotes = isSaucepan
      ? pick([s(companion['full_description']), s(v2Data['creator_notes'])])
      : pick([s(char['description']), s(v2Data['creator_notes'])]);

  // Multiple first messages ("alternate greetings"). The top-level field is
  // often an empty [] while the real list lives in chara_card_v2_json; pick the
  // first NON-EMPTY list rather than the first non-null (mirrors the reference).
  final altG = [
    char['alternate_greetings'],
    recovered?['alternate_greetings'],
    v2Data['alternate_greetings'],
  ].firstWhere(
    (a) => a is List && a.isNotEmpty,
    orElse: () => const <dynamic>[],
  );
  // Every greeting the row carries, folded into an opening line plus
  // alternates. `first_messages` is Janitor's plural field and DataCat mirrors
  // Janitor rows, so a card can arrive with its whole set there and nothing in
  // the singular field; reading only the singular one left slot one blank and
  // shifted every greeting down by one, which is the "datacat skips the 1st
  // greeting" in the report.
  final greetings = normalizeGreetings(
    primary: pick([
      s(char['first_message']),
      s(recovered?['first_message']),
      s(v2Data['first_mes']),
    ]),
    others: [
      ...greetingList(char['first_messages']),
      ...greetingList(recovered?['first_messages']),
      ...greetingList(v2Data['first_messages']),
      ...greetingList(altG),
    ],
  );

  final name = pick([s(char['chat_name']), s(char['chatName']), s(char['name'])]);

  return CharacterData(
    name: name.isEmpty ? 'Unknown' : name,
    description: definition,
    personality: '',
    scenario: scenario,
    firstMes: greetings.firstMes,
    mesExample:
        pick([s(char['example_dialogs']), s(char['mes_example']), s(v2Data['mes_example'])]),
    creatorNotes: creatorNotes,
    systemPrompt: s(v2Data['system_prompt']),
    postHistoryInstructions: s(v2Data['post_history_instructions']),
    alternateGreetings: greetings.alternates,
    tags: const [],
    creator: pick([s(char['creator_name']), s(char['creatorName'])]),
    creatorId: pick([s(char['creator_id']), s(char['creatorId'])]),
    characterBook: v2Data['character_book'],
  );
}

/// Reads the row an extraction produced.
///
/// Deliberately the plain character endpoint, NOT `/download`: `/download` is
/// gated behind a Cloudflare Turnstile lease and answers 403. The Client API
/// makes that lease obtainable and is the supported path for a character it
/// already knows — but a row this session just created is read here, in the
/// same session that created it.
Future<DownloadedCharacter> datacatGetCharacter(String uuid) async {
  final ts = DateTime.now().millisecondsSinceEpoch;
  final data = await _datacatGet('/api/characters/$uuid?t=$ts&sourceKind=janitor');
  final char = (data['character'] ?? data) as Map<String, dynamic>;
  return DownloadedCharacter(
    charData: datacatCharacterData(char),
    avatarUrl: _resolveAvatarUrl(_pickAvatarSource(char, {})),
  );
}

String _detectExtractionSource(String url) {
  if (RegExp(r'^https?://(?:www\.)?saucepan\.ai/companion/', caseSensitive: false).hasMatch(url)) {
    return 'saucepan';
  }
  return 'janitor';
}

Future<Map<String, dynamic>> _datacatExtract(String url, {bool publicFeed = true}) async {
  final idempotencyKey = _uuid();
  final source = _detectExtractionSource(url);

  if (source == 'saucepan') {
    return _datacatPost(
      '/api/saucepan-extract/run',
      {
        'companion': url,
        'extractHidden': false,
        'includeSearch': true,
        'alwaysReextract': false,
        'netnsRole': 'general_scraper',
        'sourceKind': 'one_off',
        'sourceRef': idempotencyKey,
        'vpnNamespace': 'general_scraper',
        'idempotencyKey': idempotencyKey,
      },
    );
  }

  return _datacatPost(
    '/api/character/smart-extract-v2',
    {
      'url': url,
      'appearOnPublicFeed': publicFeed,
      'useSeparateWorkerServer': true,
      'inlinePostExtractCreatorProfile': true,
      'idempotencyKey': idempotencyKey,
    },
  );
}

Future<Map<String, dynamic>> _datacatExtractionStatus() =>
    _datacatGet('/api/extraction/status?t=${DateTime.now().millisecondsSinceEpoch}');

Future<String?> datacatGetCharacterAvatar(String uuid) async {
  final ts = DateTime.now().millisecondsSinceEpoch;
  final data = await _datacatGet('/api/characters/$uuid?t=$ts');
  final char = (data['character'] ?? data) as Map<String, dynamic>;
  final meta = (data['metadata'] ?? <String, dynamic>{}) as Map<String, dynamic>;
  return _resolveAvatarUrl(_pickAvatarSource(char, meta));
}

class ExtractionResult {
  final CharacterData? charData;
  final String? avatarUrl;
  final String? characterId;
  final String? error;
  final String? phase;

  ExtractionResult({this.charData, this.avatarUrl, this.characterId, this.error, this.phase});
}

Future<ExtractionResult> datacatExtractAndPoll(
  String url, {
  void Function(String phase)? onPhaseChange,
}) async {
  try {
    final extractRes = await _datacatExtract(url);

    if (extractRes['characterId'] != null) {
      final charId = extractRes['characterId'] as String;
      final result = await datacatGetCharacter(charId);
      return ExtractionResult(
        charData: result.charData,
        avatarUrl: result.avatarUrl,
        characterId: charId,
      );
    }

    final myRequestId = extractRes['requestId'] as String?;
    final preStatus = await _datacatExtractionStatus();
    final prevRunId = (preStatus['run']?['requestId'] ?? '') as String;
    final uuidMatch = RegExp(r'[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}', caseSensitive: false)
        .firstMatch(url);
    final targetUuid = uuidMatch?.group(0);

    const maxAttempts = 60;
    // A run reported as finished-but-empty has to be seen twice before the
    // import gives up on it. The status endpoint composes the run and the
    // character it produced from different places, so a single poll can catch
    // the moment in between; three more seconds is a cheap price for not
    // calling a successful import a failure.
    var emptyReadings = 0;

    for (var attempt = 0; attempt < maxAttempts; attempt++) {
      await Future<void>.delayed(const Duration(seconds: 3));

      try {
        final status = await _datacatExtractionStatus();
        final reading = readExtractionStatus(
          status,
          requestId: myRequestId,
          previousRunId: prevRunId,
          targetUuid: targetUuid,
        );
        onPhaseChange?.call(reading.phase);

        final characterId = reading.characterId;
        if (characterId != null) {
          final result = await datacatGetCharacter(characterId);
          String? avatarUrl = result.avatarUrl;
          if (avatarUrl == null && _detectExtractionSource(url) == 'saucepan') {
            avatarUrl = await datacatGetCharacterAvatar(characterId);
          }
          return ExtractionResult(
            charData: result.charData,
            avatarUrl: avatarUrl,
            characterId: characterId,
          );
        }

        if (reading.state == ExtractionRunState.finishedEmpty) {
          if (++emptyReadings >= 2) {
            return ExtractionResult(
              error: extractionFinishedEmptyMessage(
                isSaucepan: _detectExtractionSource(url) == 'saucepan',
              ),
            );
          }
        } else {
          emptyReadings = 0;
        }
      } catch (_) {}
    }

    return ExtractionResult(error: 'Extraction timed out');
  } catch (e) {
    return ExtractionResult(error: e.toString());
  }
}
